File: projects/01-url-shortener/src/url_validate.rs
//! URL validation + normalization for submitted long URLs (security checklist).

use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};

use url::{Host, Url};

use crate::error::AppError;

/// Maximum length of a normalized URL we are willing to store, in bytes.
const MAX_URL_LEN: usize = 2048;

/// Parse, validate, and normalize a submitted long URL for storage.
///
/// Enforces the security checklist for user-supplied URLs: only `https` is
/// accepted, and hosts that point at internal infrastructure (loopback,
/// private/link-local ranges, CGNAT, `localhost`, and friends) are rejected to
/// prevent SSRF via the redirect. On success the URL is normalized — the
/// fragment is dropped and a redundant `:443` port is removed — so equivalent
/// inputs dedupe to the same stored string.
///
/// # Errors
///
/// Returns [`AppError::BadRequest`] when the input is empty, unparseable, not
/// `https`, has no host, resolves to a [blocked host](is_blocked_host), or whose
/// normalized form exceeds [`MAX_URL_LEN`].
///
/// # Examples
///
/// ```ignore
/// let url = validate_long_url("  https://example.com:443/x#frag  ").unwrap();
/// assert_eq!(url, "https://example.com/x");
///
/// assert!(validate_long_url("http://example.com").is_err());
/// assert!(validate_long_url("https://127.0.0.1/").is_err());
/// ```
pub fn validate_long_url(input: &str) -> Result<String, AppError> {
    let trimmed = input.trim();
    if trimmed.is_empty() {
        return Err(AppError::BadRequest("invalid URL".into()));
    }

    let mut url = Url::parse(trimmed).map_err(|_| AppError::BadRequest("invalid URL".into()))?;

    if !url.scheme().eq_ignore_ascii_case("https") {
        return Err(AppError::BadRequest("only https URLs are allowed".into()));
    }

    // Use the typed `Host` (not `host_str()`): an IPv6 literal stringifies with
    // brackets (`[::1]`), which fails `IpAddr::parse` and would silently skip the
    // IP checks — the exact SSRF bypass this guards against. `host()` hands back
    // an `Ipv4Addr`/`Ipv6Addr` directly. Note: this inspects only the host as
    // written; it does not resolve DNS, so a public name pointing at a private
    // address (DNS rebinding) is not caught here.
    let blocked = match url
        .host()
        .ok_or_else(|| AppError::BadRequest("invalid URL".into()))?
    {
        Host::Ipv4(ip) => is_blocked_ip(IpAddr::V4(ip)),
        Host::Ipv6(ip) => is_blocked_ip(IpAddr::V6(ip)),
        Host::Domain(name) => is_blocked_hostname(name),
    };
    if blocked {
        return Err(AppError::BadRequest("internal URLs are not allowed".into()));
    }

    // Fragment is not sent on redirect; drop it so equivalent URLs dedupe cleanly.
    url.set_fragment(None);
    if url.port() == Some(443) {
        let _ = url.set_port(None);
    }

    let normalized = url.to_string();
    if normalized.len() > MAX_URL_LEN {
        return Err(AppError::BadRequest("URL too long".into()));
    }

    Ok(normalized)
}

/// Block hostnames that conventionally resolve to local/internal endpoints.
///
/// Strips a trailing root dot and matches case-insensitively against `localhost`
/// and the `.localhost`, `.local`, and `.internal` suffixes.
fn is_blocked_hostname(host: &str) -> bool {
    let host = host.trim_end_matches('.');
    let lower = host.to_ascii_lowercase();
    lower == "localhost"
        || lower.ends_with(".localhost")
        || lower.ends_with(".local")
        || lower.ends_with(".internal")
}

/// Block IP literals that are not safely routable to the public internet.
///
/// Covers loopback, private, link-local (incl. the `169.254.169.254` cloud
/// metadata endpoint), unspecified, broadcast, the `0.0.0.0/8` range, and CGNAT
/// for IPv4; IPv6 is delegated to [`is_blocked_ipv6`].
fn is_blocked_ip(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(ip) => {
            ip.is_loopback()
                || ip.is_private()
                || ip.is_link_local()
                || ip.is_unspecified()
                || ip.is_broadcast()
                || ip.octets()[0] == 0
                || is_cgnat_ipv4(ip)
        }
        IpAddr::V6(v6) => is_blocked_ipv6(v6),
    }
}

/// 100.64.0.0/10 — shared address space (RFC 6598), often reachable internally.
fn is_cgnat_ipv4(ip: Ipv4Addr) -> bool {
    let [a, b, _, _] = ip.octets();
    a == 100 && (b & 0xC0) == 64
}

/// Block IPv6 literals that target the local host or an internal network.
fn is_blocked_ipv6(ip: Ipv6Addr) -> bool {
    ip.is_loopback() || ip.is_unspecified() || is_unique_local_ipv6(ip) || is_link_local_ipv6(ip)
}

/// Match the IPv6 unique-local range `fc00::/7` (RFC 4193), the IPv6 analogue of
/// private address space.
fn is_unique_local_ipv6(ip: Ipv6Addr) -> bool {
    (ip.segments()[0] & 0xFE00) == 0xFC00
}

/// Match the IPv6 link-local range `fe80::/10` (RFC 4291).
fn is_link_local_ipv6(ip: Ipv6Addr) -> bool {
    (ip.segments()[0] & 0xFFC0) == 0xFE80
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_public_https_url() {
        let url = validate_long_url("https://example.com/path?q=1").unwrap();
        assert_eq!(url, "https://example.com/path?q=1");
    }

    #[test]
    fn trims_whitespace() {
        let url = validate_long_url("  https://example.com  ").unwrap();
        assert_eq!(url, "https://example.com/");
    }

    #[test]
    fn rejects_http_scheme() {
        assert!(validate_long_url("http://example.com").is_err());
    }

    #[test]
    fn rejects_javascript_scheme() {
        assert!(validate_long_url("javascript:alert(1)").is_err());
    }

    #[test]
    fn rejects_loopback_ipv4() {
        assert!(validate_long_url("https://127.0.0.1/").is_err());
    }

    #[test]
    fn rejects_private_ipv4() {
        assert!(validate_long_url("https://192.168.1.1/").is_err());
        assert!(validate_long_url("https://10.0.0.1/").is_err());
    }

    #[test]
    fn rejects_link_local_metadata_ip() {
        assert!(validate_long_url("https://169.254.169.254/").is_err());
    }

    #[test]
    fn rejects_localhost_hostname() {
        assert!(validate_long_url("https://localhost/").is_err());
    }

    #[test]
    fn strips_fragment_and_default_https_port() {
        let url = validate_long_url("https://example.com:443/x#section").unwrap();
        assert_eq!(url, "https://example.com/x");
    }

    #[test]
    fn allows_http_in_query_string() {
        let url = validate_long_url("https://example.com?next=http://other.example").unwrap();
        assert_eq!(url, "https://example.com/?next=http://other.example");
    }

    #[test]
    fn rejects_over_length_url() {
        // Otherwise valid (https, public host); only the length rule can reject it.
        let long = format!("https://example.com/{}", "a".repeat(MAX_URL_LEN));
        assert!(long.len() > MAX_URL_LEN);
        assert!(validate_long_url(&long).is_err());
    }

    #[test]
    fn accepts_url_at_the_length_limit() {
        // Boundary: exactly MAX_URL_LEN bytes must still be accepted, proving the
        // check rejects *over*-length, not at-length.
        let path_len = MAX_URL_LEN - "https://example.com/".len();
        let at_limit = format!("https://example.com/{}", "a".repeat(path_len));
        assert_eq!(at_limit.len(), MAX_URL_LEN);
        assert!(validate_long_url(&at_limit).is_ok());
    }

    #[test]
    fn rejects_loopback_and_ula_ipv6() {
        // IPv6 SSRF: loopback (::1) and unique-local (fc00::/7). Bracketed in the
        // authority per RFC 3986.
        assert!(validate_long_url("https://[::1]/").is_err());
        assert!(validate_long_url("https://[fc00::1]/").is_err());
        assert!(validate_long_url("https://[fe80::1]/").is_err());
    }
}

File: projects/01-url-shortener/src/url_shortener/url_validate.py
"""URL validation + normalization for submitted long URLs (security checklist).

A shortener is an open redirector by construction: whatever you store here, the
service will later hand to a browser as a `Location`. That makes the create path
an SSRF surface - someone shortens `https://169.254.169.254/latest/meta-data/`
and then gets *your users* (or your own link-preview crawler) to fetch it. So the
host is inspected before anything is stored.

What this does **not** do is resolve DNS. Only the host as written is checked, so
a public name that resolves to a private address (DNS rebinding) still gets
through. Catching that means resolving at fetch time and pinning the address,
which is a different piece of machinery than a validator.
"""

from __future__ import annotations

import ipaddress
from urllib.parse import urlsplit, urlunsplit

from .errors import BadRequest

__all__ = ["MAX_URL_LEN", "validate_long_url"]

MAX_URL_LEN = 2048
"""Longest normalized URL we are willing to store, in bytes."""

_BLOCKED_HOST_SUFFIXES = (".localhost", ".local", ".internal")


def validate_long_url(raw: str) -> str:
    """Parse, validate, and normalize a submitted long URL for storage.

    Only `https` is accepted, and hosts that point at internal infrastructure are
    rejected. On success the URL is normalized - the fragment is dropped and a
    redundant `:443` removed - so equivalent inputs dedupe to the same string.

    Raises:
        BadRequest: empty, unparseable, not https, hostless, internal, or longer
            than :data:`MAX_URL_LEN` once normalized.
    """
    trimmed = raw.strip()
    if not trimmed:
        raise BadRequest("invalid URL")

    try:
        parts = urlsplit(trimmed)
    except ValueError as exc:  # malformed IPv6 literal, bad port, ...
        raise BadRequest("invalid URL") from exc

    if parts.scheme.lower() != "https":
        raise BadRequest("only https URLs are allowed")

    try:
        # `.hostname` is the reason this reads so plainly: it lower-cases the
        # host and *strips the brackets* off an IPv6 literal. Using the raw
        # netloc instead would hand `[::1]` to the IP parser, which rejects it -
        # and a rejected parse would fall through to the hostname rules and pass.
        # That is the exact SSRF bypass this guards against.
        hostname = parts.hostname
        port = parts.port
    except ValueError as exc:
        raise BadRequest("invalid URL") from exc

    if not hostname:
        raise BadRequest("invalid URL")

    if _is_blocked_host(hostname):
        raise BadRequest("internal URLs are not allowed")

    # Fragment is never sent on a redirect; drop it so equivalent URLs dedupe.
    netloc = parts.netloc
    if port == 443:
        netloc = _strip_default_port(parts.netloc)
    normalized = urlunsplit((parts.scheme.lower(), netloc, parts.path or "/", parts.query, ""))

    if len(normalized) > MAX_URL_LEN:
        raise BadRequest("URL too long")
    return normalized


def _strip_default_port(netloc: str) -> str:
    """Remove a trailing `:443` from an authority, IPv6 brackets intact."""
    head, _, _ = netloc.rpartition(":")
    return head


def _is_blocked_host(hostname: str) -> bool:
    """True when the host is not safely routable to the public internet."""
    try:
        ip = ipaddress.ip_address(hostname)
    except ValueError:
        return _is_blocked_hostname(hostname)

    # `::ffff:127.0.0.1` is a loopback address wearing an IPv6 costume, and it is
    # "global" as far as the v6 rules are concerned. Unwrap before judging.
    if isinstance(ip, ipaddress.IPv6Address) and ip.ipv4_mapped is not None:
        ip = ip.ipv4_mapped

    # One question instead of a list of ranges. `is_global` is False for
    # loopback, private, link-local (incl. the 169.254.169.254 cloud metadata
    # endpoint), unspecified, reserved, multicast, and CGNAT (100.64.0.0/10) -
    # and it stays correct as the registries change, which a hand-rolled
    # enumeration does not.
    return not ip.is_global


def _is_blocked_hostname(hostname: str) -> bool:
    """Block names that conventionally resolve to local/internal endpoints."""
    host = hostname.rstrip(".").lower()
    return host == "localhost" or host.endswith(_BLOCKED_HOST_SUFFIXES)
